services/gitdiff/gitdiff.go

```diff
@@ -1109,12 +1100,19 @@ type DiffOptions struct {
 func GetDiff(ctx context.Context, gitRepo *git.Repository, opts *DiffOptions, files ...string) (*Diff, error) {
 	repoPath := gitRepo.Path
 
-	commit, err := gitRepo.GetCommit(opts.AfterCommitID)
-	if err != nil {
-		return nil, err
+	commit := opts.AfterCommit
+	if commit == nil {
+		var err error
+		commit, err = gitRepo.GetCommit(opts.AfterCommitID)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	cmdDiff := git.NewCommand(gitRepo.Ctx)
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
It shouldn't override the old "ctx". Otherwise, if there are other accesses to "ctx", they would get a canceled ctx and fail the following operations.

Such bug has happened in history, very difficult to debug.

So ideally, it should be `cmdDiffCtx, cmdDiffCancel := context.WithCancel(ctx)`

Reply:
This is a different case. Here, the ctx will be canceled soon. It's like this:

```
ctx, cancel := ...
useCtx(ctx)
cancel()

// if there are other accesses to ctx (in the future)
anotherUseCtx(ctx) // it won't work because the ctx has been canceled above.
```

So, if you are introducing a local ctx which will be canceled soon, it shouldn't override the parent ctx's name.

Reply:
fixed in https://github.com/go-gitea/gitea/pull/32393/commits/d8e1e41aaed7e6e55f9606725be7d0264bbd7235